File: hw-3/src/lexer.rs
use regex::Regex;

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum LiteralValue {
    String(String),
    Number(i64),
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Token {
    Comment(String),
    Set,
    ArrayBegin,
    ArrayEnd,
    DictBegin,
    DictSep,
    DictEnd,
    Identifier(String),
    ExprBegin,
    ExprEnd,
    Delimiter,
    Literal(LiteralValue),
}

static SIMPLE_TOKENS: [(&'static str, Token); 9] = [
    ("set", Token::Set),
    ("array(", Token::ArrayBegin),
    (")", Token::ArrayEnd),
    ("{", Token::DictBegin),
    (":", Token::DictSep),
    ("}", Token::DictEnd),
    ("?[", Token::ExprBegin),
    ("]", Token::ExprEnd),
    (",", Token::Delimiter),
];

fn lex_simple(text: &str) -> Option<(Token, &str)> {
    let (str, token) = SIMPLE_TOKENS
        .iter()
        .find(|token| text.starts_with(token.0))?;

    return Some((token.clone(), &text[str.len()..]));
}

fn lex_comment(text: &str) -> Option<(Token, &str)> {
    if !text.starts_with("#=") {
        return None;
    }

    let end = text.find("=#")?;
    let comment = &text[2..end];

    Some((Token::Comment(comment.into()), &text[end + 2..]))
}

fn lex_number(text: &str) -> Option<(Token, &str)> {
    let digits: String = text
        .chars()
        .take_while(|chr| char::is_numeric(*chr))
        .collect();

    if digits.len() == 0 {
        return None;
    }

    Some((
        Token::Literal(LiteralValue::Number(digits.parse().ok()?)),
        &text[digits.len()..],
    ))
}

fn lex_string(text: &str) -> Option<(Token, &str)> {
    if text.chars().next()? != '"' {
        return None;
    }

    let end = text[1..].find('"')? + 1;

    return Some((
        Token::Literal(LiteralValue::String(text[1..end].into())),
        &text[end + 1..],
    ));
}

fn lex_identifier(text: &str) -> Option<(Token, &str)> {
    let regex = Regex::new("^[a-zA-Z][_a-zA-Z0-9]*").unwrap();
    let identifier = regex.find(text)?;

    return Some((
        Token::Identifier(text[..identifier.end()].into()),
        &text[identifier.end()..],
    ));
}

pub fn lex(text: &str) -> Option<Vec<Token>> {
    let mut result: Vec<Token> = Vec::new();
    let mut text = text;

    loop {
        text = text.trim_start();
        let token = None
            .or_else(|| lex_simple(text))
            .or_else(|| lex_comment(text))
            .or_else(|| lex_number(text))
            .or_else(|| lex_string(text));

        if let Some((token, rest)) = token {
            text = rest;
            result.push(token);
            continue;
        }

        break;
    }

    if text != "" {
        return None;
    }

    return Some(result);
}

File: hw-3/app/Lexer.hs
module Lexer
  ( LiteralValue (..),
    Token (..),
    LexingError (..),
    tokenize,
  )
where

import Control.Monad ((>=>))
import Data.Char (isDigit, isSpace)
import Data.List (stripPrefix)
import Text.Read (readMaybe)
import Text.Regex.TDFA ((=~))
import Util (maybeToEither, splitOnce)

data LiteralValue
  = String String
  | Number Integer
  deriving (Show, Eq)

data Token
  = Set
  | Equals
  | ArrayBegin
  | ArrayEnd
  | DictBegin
  | DictSep
  | DictEnd
  | Delimiter
  | ExprBegin
  | Add
  | Subtract
  | Multiply
  | Print
  | ExprEnd
  | Comment String
  | Identifier String
  | Literal LiteralValue
  deriving (Show, Eq)

newtype LexingError
  = NoValidToken String
  deriving (Show)

newtype Lexer = Lexer (String -> Either LexingError (Token, String))

instance Semigroup Lexer where
  (Lexer left) <> (Lexer right) = Lexer $ \text -> case left text of
    Right res -> Right res
    _ -> right text

instance Monoid Lexer where
  mempty = Lexer $ \text -> Left $ NoValidToken text

stripPrefix' :: String -> String -> Either LexingError String
stripPrefix' pref text = maybeToEither (NoValidToken text) $ stripPrefix pref text

splitOnce' :: String -> String -> Either LexingError (String, String)
splitOnce' delim text = maybeToEither (NoValidToken text) $ splitOnce delim text

lexSimple :: String -> Token -> Lexer
lexSimple prefix token = Lexer $ stripPrefix' prefix >=> (\rest -> Right (token, rest))

lexComment :: Lexer
lexComment = Lexer $ \text -> do
  rest <- stripPrefix' "#=" text
  (comment, rest') <- splitOnce' "=#" rest
  return (Comment comment, rest')

lexIdentifier :: Lexer
lexIdentifier = Lexer $ \text -> do
  identifier <- case (text =~ "^[a-zA-Z][_a-zA-Z0-9]*" :: String) of
    "" -> Left $ NoValidToken text
    str -> Right str

  rest <- stripPrefix' identifier text
  return (Identifier identifier, rest)

lexNumber :: Lexer
lexNumber = Lexer $ \text -> do
  let digits = takeWhile isDigit text
  number <- maybeToEither (NoValidToken text) $ readMaybe digits
  rest <- stripPrefix' digits text
  return (Literal (Number number), rest)

lexString :: Lexer
lexString = Lexer $ \text -> do
  rest <- stripPrefix' "\"" text
  (string, rest') <- splitOnce' "\"" rest
  return (Literal (String string), rest')

nextToken :: String -> Either LexingError (Token, String)
Lexer nextToken =
  mconcat
    [ lexSimple "set" Set,
      lexSimple "=" Equals,
      lexSimple "array(" ArrayBegin,
      lexSimple ")" ArrayEnd,
      lexSimple "{" DictBegin,
      lexSimple ":" DictSep,
      lexSimple "}" DictEnd,
      lexSimple "," Delimiter,
      lexSimple "?[" ExprBegin,
      lexSimple "+" Add,
      lexSimple "-" Subtract,
      lexSimple "*" Multiply,
      lexSimple "print" Print,
      lexSimple "]" ExprEnd,
      lexComment,
      lexIdentifier,
      lexNumber,
      lexString
    ]

tokenize :: String -> Either LexingError [Token]
tokenize "" = Right []
tokenize text = do
  (token, rest) <- nextToken (dropWhile isSpace text)
  tokens <- tokenize rest
  return (token : tokens)
